### PR TITLE
Change pod domain name pattern to omit ClusterDomain if empty

### LINKED ISSUE #121868: Kubelet can generate an invalid fully qualified domain name for a pod
Kubelet can generate an invalid fully qualified domain name for a pod when the `ClusterDomain` configured in the kubelets config.yaml defaults to ""

The code in function `GeneratePodHostNameAndDomain` can result in a fully qualified domain name ending in a period, which is invalid.

The problematic line is `hostDomain = fmt.Sprintf("%s.%s.svc.%s", pod.Spec.Subdomain, pod.Namespace, clusterDomain)`

https://github.com/kubernetes/kubernetes/blob/36930f4cb605450ccbe3126caad365aa5f51267c/pkg/kubelet/kubelet_pods.go#L455

The code already has a check for length of subdomain string being > 0.  I wonder if a similar check for the ClusterDomain string would be sensible?

Note:  I believe that kubeadm will set `ClusterDomain` in the kubelets config to a non empty string, but it remains the case that the default of "" doesn't get handled well.  Id be happy to submit a fix, but would appreciate insights from an SME first.

### PR DESCRIPTION
**Pull request by vlasebian**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

During the creation of the fully qualified domain name for a pod, the domain name might end in '.' if the ClusterDomain is empty. This PR makes sure to append the ClusterDomain only if it is not empty, thus making sure the domain name for the pod will not end in a '.'.

#### Which issue(s) this PR fixes:

Fixes #121868 

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/kubelet/kubelet_pods.go
Comment (SergeyKanzhelev) [inline pkg/kubelet/kubelet_pods.go:453]: can you please add a unit test for this method? Code change lgtm, but it will be best if it will be covered with the unit tests

Comment (vlasebian) [reply pkg/kubelet/kubelet_pods.go:453]: Sure! I just added the unit test, let me know if there is anything else missing.

File: pkg/kubelet/kubelet_pods_test.go
Comment (HirazawaUi) [inline pkg/kubelet/kubelet_pods_test.go:6062]:
```suggestion
	testCases := []struct {
		name             string
		podSpecHostname  string
		podSpecSubdomain string
		clusterDomain    string
		wantHostname     string
		wantHostDomain   string
		mustErr          bool
	}
```

File: pkg/kubelet/kubelet_pods_test.go
Comment (HirazawaUi) [inline pkg/kubelet/kubelet_pods_test.go:6075]: nit: unnecessary blank lines

File: pkg/kubelet/kubelet_pods_test.go
Comment (HirazawaUi) [inline pkg/kubelet/kubelet_pods_test.go:6085]: nit: unnecessary blank lines

File: pkg/kubelet/kubelet_pods_test.go
Comment (vlasebian) [reply pkg/kubelet/kubelet_pods_test.go]: fixed

File: pkg/kubelet/kubelet_pods_test.go
Comment (vlasebian) [reply pkg/kubelet/kubelet_pods_test.go]: fixed

File: pkg/kubelet/kubelet_pods_test.go
Comment (vlasebian) [reply pkg/kubelet/kubelet_pods_test.go]: fixed

**Comment by HirazawaUi** [comment]
It looks like the `pull-kubernetes-node-e2e-containerd` error is related to your changes, can you try to fix it?

**Comment by vlasebian** [comment]
yes, on it!

**Comment by vlasebian** [comment]
Hi, @HirazawaUi! I checked the errors, there are 2 tests that fail related to this change, which ran twice. For the fifth test, I'm not sure yet if it was caused by my change or not, I will investigate more.

1. E2eNode Suite: [It] [sig-node] Hostname of Pod [NodeConformance] a pod with subdomain field has FQDN, hostname is shortname

The expected pod FQDN is constructed as:
```
# ./test/e2e_node/pod_hostnamefqdn_test.go
121: // Expected Pod FQDN
122: hostFQDN := fmt.Sprintf("%s.%s.%s.svc.%s", pod.ObjectMeta.Name, subdomain, f.Namespace.Name, framework.TestContext.ClusterDNSDomain)
````
In case the clusterDNSDomain is empty, the expected hostFQDN will end with a `.`, but the actual result won't have it anymore.

2. E2eNode Suite: [It] [sig-node] Hostname of Pod [NodeConformance] a pod with subdomain field has FQDN, when setHostnameAsFQDN is set to true, the FQDN is set as hostname 

In this test, same case as above:

```
# ./test/e2e_node/pod_hostnamefqdn_test.go
143: // Expected Pod FQDN
144: hostFQDN := fmt.Sprintf("%s.%s.%s.svc.%s", pod.ObjectMeta.Name, subdomain, f.Namespace.Name, framework.TestContext.ClusterDNSDomain)
```

The two tests that fail are Node Conformance tests. I didn't change them yet because I don't know if it's appropriate to do this in this PR. Is it okay to do them here or should I open another PR? Also, should someone be noticed of this change beforehand as these are node conformance tests?

**Comment by HirazawaUi** [comment]
> I don't know if it's appropriate to do this in this PR. Is it okay to do them here or should I open another PR?

You can add a new commit to the current PR to fix these test errors.

File: test/e2e_node/pod_hostnamefqdn_test.go
Comment (bart0sh) [inline test/e2e_node/pod_hostnamefqdn_test.go:150]: Would it make sense to have a helper function to avoid code duplication?

**Review by SergeyKanzhelev — APPROVED** [review]
/lgtm
/approve

File: test/e2e_node/pod_hostnamefqdn_test.go
Comment (vlasebian) [reply test/e2e_node/pod_hostnamefqdn_test.go:122]: Yes, the tests started failing.

**Comment by aojea** [comment]
/hold


all fully-qualified domain names end with a dot, I'd like to understand better the problems before start PRs, can we please describe better the problems caused by this https://github.com/kubernetes/kubernetes/issues/121868
?

**Comment by bart0sh** [comment]
@aojea 
> all fully-qualified domain names end with a dot

Can you point out to some spec/examples?

[RFC 952](http://www.faqs.org/rfcs/rfc952.html) states that
   A "name" (Net, Host, Gateway, or Domain name) is a text string up
   to 24 characters drawn from the alphabet (A-Z), digits (0-9), minus
   sign (-), and period (.).  **Note that periods are only allowed when
   they serve to delimit components of "domain style names".** (See
   [RFC-921](http://www.faqs.org/rfcs/rfc921.html), "Domain Name System Implementation Schedule", for
   background).  No blank or space characters are permitted as part of a
   name. No distinction is made between upper and lower case.  The first
   character must be an alpha character.  **The last character must not be
   a minus sign or period.**  A host which serves as a GATEWAY should have
   "-GATEWAY" or "-GW" as part of its name.  Hosts which do not serve as
   Internet gateways should not use "-GATEWAY" and "-GW" as part of
   their names. A host which is a TAC should have "-TAC" as the last
   part of its host name, if it is a DoD host.  Single character names
   or nicknames are not allowed.

**Comment by vlasebian** [comment]
I'm closing the PR because the issue it was intended for was closed.

**Comment by aojea** [comment]
@bart0sh https://github.com/kubernetes/kubernetes/issues/121868#issuecomment-1875535208

**Pull request closed without merging**
